be/src/storage/lake/lake_persistent_index.cpp

```diff
@@ -175,8 +218,109 @@ Status LakePersistentIndex::try_replace(size_t n, const Slice* keys, const Index
+    for (const auto& sstable : _sstables) {
+        sstable::Iterator* iter = sstable->new_iterator(read_options);
+        iters.emplace_back(iter);
+        if (iters.size() >= config::lake_pk_index_sst_max_compaction_versions) {
+            break;
+        }
+    }
+    sstable::Options options;
+    sstable::Iterator* iter = sstable::NewMergingIterator(options.comparator, &iters[0], iters.size());
+    std::unique_ptr<sstable::Iterator> iter_ptr = nullptr;
+    iter_ptr.reset(iter);
+    iter_ptr->SeekToFirst();
+    return iter_ptr;
+}
+
+Status LakePersistentIndex::merge_sstables(std::unique_ptr<sstable::Iterator> iter_ptr,
+                                           sstable::TableBuilder* builder) {
+    auto merger = std::make_unique<KeyValueMerger>(iter_ptr->key().to_string(), builder);
+    while (iter_ptr->Valid()) {
+        RETURN_IF_ERROR(merger->merge(iter_ptr->key().to_string(), iter_ptr->value().to_string()));
+        iter_ptr->Next();
+    }
+    RETURN_IF_ERROR(iter_ptr->status());
+    merger->finish();
+    return builder->Finish();
+}
+
+Status LakePersistentIndex::major_compact(int64_t min_retain_version, std::shared_ptr<TxnLogPB>& txn_log) {
+    if (_sstables.size() < 2) {
+        return Status::OK();
+    }
+
+    auto iter_ptr = prepare_merging_iterator();
+    if (!iter_ptr->Valid()) {
+        return iter_ptr->status();
+    }
+
+    auto filename = gen_sst_filename();
+    auto location = _tablet_mgr->sst_location(_tablet_id, filename);
+    ASSIGN_OR_RETURN(auto wf, fs::new_writable_file(location));
+    sstable::Options options;
+    std::unique_ptr<sstable::FilterPolicy> filter_policy;
+    filter_policy.reset(const_cast<sstable::FilterPolicy*>(sstable::NewBloomFilterPolicy(10)));
+    options.filter_policy = filter_policy.get();
+    sstable::TableBuilder builder(options, wf.get());
+    RETURN_IF_ERROR(merge_sstables(std::move(iter_ptr), &builder));
+    RETURN_IF_ERROR(wf->close());
+
+    for (const auto& sstable : _sstables) {
+        auto input_sstable = txn_log->mutable_op_compaction()->add_input_sstables();
+        auto sstable_pb = sstable->sstable_pb();
+        input_sstable->CopyFrom(sstable_pb);
+        if (txn_log->op_compaction().input_sstables_size() >= config::lake_pk_index_sst_max_compaction_versions) {
+            break;
+        }
+    }
+    txn_log->mutable_op_compaction()->mutable_output_sstable()->set_filename(filename);
+    txn_log->mutable_op_compaction()->mutable_output_sstable()->set_filesize(builder.FileSize());
     return Status::OK();
 }
 
+Status LakePersistentIndex::apply_opcompaction(const TxnLogPB_OpCompaction& op_compaction) {
+    if (op_compaction.input_sstables().empty()) {
+        return Status::OK();
+    }
+
+    std::unordered_set<std::string> filenames;
+    for (const auto& input_sstable : op_compaction.input_sstables()) {
+        filenames.insert(input_sstable.filename());
+    }
+    for (auto it = _sstables.begin(); it != _sstables.end();) {
```

Comment:
if auto increasing is not leveraged, while(xx) loop might be better than for loop.